trino-s3-proxy/src/test/java/io/trino/s3/proxy/server/testing/ManagedS3MockContainer.java

```diff
@@ -28,24 +33,58 @@
 
 public class ManagedS3MockContainer
 {
-    private final S3MockContainer container;
+    private final MinIOContainer container;
+
+    private static final String CONFIG = """
+            {
+                "version": "10",
+                "aliases": {
+                    "local": {
+                        "url": "http://localhost:9000",
+                        "accessKey": "%s",
+                        "secretKey": "%s",
+                        "api": "S3v4",
+                        "path": "auto"
+                    }
+                }
+            }
+            """;
 
     @Retention(RUNTIME)
     @Target({FIELD, PARAMETER, METHOD})
     @BindingAnnotation
     public @interface ForS3MockContainer {}
 
     @Inject
-    public ManagedS3MockContainer(@ForS3MockContainer String initialBuckets)
+    public ManagedS3MockContainer(@ForS3MockContainer String initialBuckets, @ForS3MockContainer Credential credential)
     {
-        container = new S3MockContainer("3.8.0");
+        Transferable transferable = Transferable.of(CONFIG.formatted(credential.accessKey(), credential.secretKey()));
+        container = new MinIOContainer("minio/minio:RELEASE.2023-09-04T19-57-37Z")
+                .withUserName(credential.accessKey())
+                .withPassword(credential.secretKey())
+                // setting this allows us to shell into the container and run "mc" commands
+                .withCopyToContainer(transferable, "/root/.mc/config.json");
+
+        container.configure();
+        container.start();
+
         if (!initialBuckets.isEmpty()) {
-            container.withInitialBuckets(initialBuckets);
+            Splitter.on(',').trimResults().splitToStream(initialBuckets).forEach(bucket -> {
```

Comment:
If the constructor fails, `container.stop` will never be called, as the constructor of this class never terminated.

While testcontainers uses ryuk to prevent this, we could also do our bit to help if we use a try-with-resources? wdyt?

```java
try (container) {
    if (!initialBuckets.isEmpty()) {
        // ...
    }
}
```

Reply:
I moved the initial bucket setup into a `@PostConstruct` method. That will solve it.